component/otelcol/processor/discovery/discovery.go

```diff
@@ -0,0 +1,151 @@
+	_ river.Validator = (*Arguments)(nil)
+)
+
+// DefaultArguments holds default settings for Arguments.
+var DefaultArguments = Arguments{
+	OperationType: promsdconsumer.OperationTypeUpsert,
+	PodAssociations: []string{
+		promsdconsumer.PodAssociationIPLabel,
+		promsdconsumer.PodAssociationOTelIPLabel,
+		promsdconsumer.PodAssociationk8sIPLabel,
+		promsdconsumer.PodAssociationHostnameLabel,
+		promsdconsumer.PodAssociationConnectionIP,
+	},
+}
+
+// SetToDefault implements river.Defaulter.
+func (args *Arguments) SetToDefault() {
+	*args = DefaultArguments
+}
+
+// Validate implements river.Validator.
+func (args *Arguments) Validate() error {
+	err := promsdconsumer.ValidateOperationType(args.OperationType)
+	if err != nil {
+		return err
+	}
+
+	err = promsdconsumer.ValidatePodAssociations(args.PodAssociations)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Component is the otelcol.exporter.discovery component.
+type Component struct {
+	cfg      Arguments
+	consumer *promsdconsumer.Consumer
+	logger   log.Logger
+}
+
+var _ component.Component = (*Component)(nil)
+
+// New creates a new otelcol.exporter.discovery component.
+func New(o component.Options, c Arguments) (*Component, error) {
+	if c.Output.Logs != nil || c.Output.Metrics != nil {
+		level.Warn(o.Logger).Log("msg", "non-trace output detected; this component only works for traces")
+	}
+
+	nextTraces := fanoutconsumer.Traces(c.Output.Traces)
+	consumer, err := promsdconsumer.NewConsumer(nextTraces, c.OperationType, c.PodAssociations, o.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a traces consumer due to error: %w", err)
+	}
+
+	res := &Component{
+		consumer: consumer,
+		logger:   o.Logger,
+	}
+
+	if err := res.Update(c); err != nil {
+		return nil, err
+	}
+
+	// Export the consumer.
+	// This will remain the same throughout the component's lifetime,
+	// so we do this during component construction.
+	export := lazyconsumer.New(context.Background())
+	export.SetConsumers(res.consumer, nil, nil)
+	o.OnStateChange(otelcol.ConsumerExports{Input: export})
+
+	return res, nil
+}
+
+// Run implements Component.
+func (c *Component) Run(ctx context.Context) error {
+	for range ctx.Done() {
+		return nil
+	}
+	return nil
+}
+
+// Update implements Component.
+func (c *Component) Update(newConfig component.Arguments) error {
+	cfg := newConfig.(Arguments)
+	c.cfg = cfg
+
+	hostLabels := make(map[string]discovery.Target)
+
+	for _, labels := range c.cfg.Targets {
+		host, err := promsdconsumer.GetHostFromLabels(labels)
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "ignoring target, unable to find address", "err", err)
+			continue
+		}
+		promsdconsumer.CleanupLabels(labels)
+		hostLabels[host] = labels
+	}
+
+	c.consumer.UpdateOptions(promsdconsumer.Options{
+		HostLabels:      hostLabels,
+		OperationType:   cfg.OperationType,
+		PodAssociations: cfg.PodAssociations,
+		NextConsumer:    fanoutconsumer.Traces(cfg.Output.Traces),
```

Comment:
@mattdurham Apologies, but I had push a commit to fix another issue after the latest review. I wasn't updating the "next consumer".